general/bundler_imports.md

```diff
@@ -0,0 +1,160 @@
+# Bundler Imports
+
+This document is aimed at explaining how npm bundler imports work.
+
+## Why do we need npm bundler imports
+
+As all of you probably know, usually JavaScript web applications need a bundling phase to convert npm modules into a single `.js` file that can be loaded and executed by browsers. The reference tool for this task is [webpack](https://webpack.js.org/).
+
+Sadly, in our DXP platform, we cannot simply create bundles for every portlet and deploy them (or to be more specific: we could, but it would be far from optimal). The main reason being that because our HTML pages are assembled dinamically during runtime (due to portlet layouts being totally configurable), different portlets may need to share the same JavaScript modules.
+
+So, imagine you deploy two portlets (`A` and `B`) and both use `react`. You could create a bundle for each with `webpack`, but that would put two different copies of `react` in the browser memory and you would spend double the bandwidth and processing time fetching and parsing them.
+
+The ideal solution would be making both portlets share the same `react` copy. This way not only you optimize bandwidth, CPU and RAM, but also make sure that both portlets call the same `react` thus minimizing the possibility of issues arising when mixing different versions of the same framework in one single HTML page.
+
+So, how do you do that? That's when npm bundler imports come to the rescue!
+
+## What are npm bundler imports
+
+> Simply put, npm bundler imports are a way of diverting standard npm module imports from one portlet to another.
+
+As we said above, if you deployed two portlets `A` and `B` using `webpack`, each one would have its own copy of `react`. By default, our npm bundler, mimicks that behavior but, because we don't bundle things in a single `.js` file, but deploy them as AMD modules, it namespaces each `react` with the name of its owner portlet, thus, without configuring anything else, you would deploy `A$react` and `B$react`.
+
+### So, how do we change this into the ideal situation (one single `react` to rule them all)?
+
+If you look at `DXP`'s source tree, there's one project called `frontend-js-react-web`. This one is responsible for providing `react` to the rest of the `DXP` portlets.
+
+Because this portlet is bundled with our npm bundler, it publishes a copy of `react` under the name `frontend-js-react-web$react` to the AMD loader.
+
+Now, what we want is `A` and `B` to use that `frontend-js-react-web$react`, instead of `A$react` and `B$react`. And for that we will use npm bundler imports.
+
+### How npm bundler imports are configured
+
+The way to configure npm bundler imports is straighforward: if you want to use `react` from `frontend-js-react-web` just say so in your `.npmbundlerrc` file like, for example, this:
```

Comment:
> The way to configure npm bundler imports is straighforward:

This is a _lie_ and should be deleted!

Just kidding. But it should still be deleted. Less is more and all that.

Reply:
> The way to configure npm bundler imports is straighforward:

I was referring to the structure of the configuration, not to the task itself, which I agree is harder than reading the future in coffee grounds...

I will rephrase that to avoid people suffering after being disappointed.